users/vieting/experiments/switchboard/hybrid/feat/specaug_random.py

```diff
@@ -0,0 +1,152 @@
+    cond = tf.logical_and(tf.greater_equal(idxs, pos_bc), tf.less(idxs, pos2_bc))  # (batch,dim)
+    if batch_axis > axis:
+        cond = tf.transpose(cond)  # (dim,batch)
+    cond = tf.reshape(cond, [tf.shape(x)[i] if i in (batch_axis, axis) else 1 for i in range(ndim)])
+    from TFUtil import where_bc
+
+    x = where_bc(cond, 0.0, x)
+    return x
+
+
+def _random_mask(x, batch_axis, axis, min_num, max_num, max_dims):
+    """
+    :param tf.Tensor x: (batch,time,feature)
+    :param int batch_axis:
+    :param int axis:
+    :param int|tf.Tensor min_num:
+    :param int|tf.Tensor max_num: inclusive
+    :param int|tf.Tensor max_dims: inclusive
+    """
+    from returnn.tf.compat import v1 as tf
+
+    n_batch = tf.shape(x)[batch_axis]
+    if isinstance(min_num, int) and isinstance(max_num, int) and min_num == max_num:
+        num = min_num
+    else:
+        num = tf.random_uniform(shape=(n_batch,), minval=min_num, maxval=max_num + 1, dtype=tf.int32)
+    # https://github.com/tensorflow/tensorflow/issues/9260
+    # https://timvieira.github.io/blog/post/2014/08/01/gumbel-max-trick-and-weighted-reservoir-sampling/
+    z = -tf.log(-tf.log(tf.random_uniform((n_batch, tf.shape(x)[axis]), 0, 1)))
+    _, indices = tf.nn.top_k(z, num if isinstance(num, int) else tf.reduce_max(num))
+    # indices should be sorted, and of shape (batch,num), entries (int32) in [0,dim)
+    # indices = tf.Print(indices, ["indices", indices, tf.shape(indices)])
+    if isinstance(num, int):
+        for i in range(num):
+            x = _mask(x, batch_axis=batch_axis, axis=axis, pos=indices[:, i], max_amount=max_dims)
+    else:
+        _, x = tf.while_loop(
+            cond=lambda i, _: tf.less(i, tf.reduce_max(num)),
+            body=lambda i, x: (
+                i + 1,
+                tf.where(
+                    tf.less(i, num),
+                    _mask(x, batch_axis=batch_axis, axis=axis, pos=indices[:, i], max_amount=max_dims),
+                    x,
+                ),
+            ),
+            loop_vars=(0, x),
+        )
+    return x
+
+
+def specaugment_eval_func(data, network, time_factor=1):
+    x = data.placeholder
+    from returnn.tf.compat import v1 as tf
+
+    # summary("features", x)
+    step = network.global_train_step
+    step1 = tf.where(tf.greater_equal(step, 1000), 1, 0)
+    step2 = tf.where(tf.greater_equal(step, 2000), 1, 0)
+
+    def get_masked():
+        x_masked = x
+        x_masked = _random_mask(
+            x_masked,
+            batch_axis=data.batch_dim_axis,
+            axis=data.time_dim_axis,
+            min_num=step1 + step2,
+            max_num=tf.maximum(tf.shape(x)[data.time_dim_axis] // 100, 2) * (1 + step1 + step2 * 2),
+            max_dims=20 // time_factor,
+        )
+        x_masked = _random_mask_mixed(
+            x_masked,
+            batch_axis=data.batch_dim_axis,
+            axis=data.feature_dim_axis,
+            amount=(1 + step1 + step2 * 2) * (data.dim // 5),
```

Comment:
Before, we had `min_num=step1 + step2`, `max_num=2 + step1 + step2 * 2` and `max_dims=data.dim // 5`. Can you please explain how you transferred that into the equation above?

Reply:
Since we do not have to choose a number of masks I selected a value one smaller than the max value from before. This is indeed a little bit arbitrary. Alternatively, we could randomize the amount of masked elements again between `(step1 + step2)*(max_dims=data.dim // 5)` and `(2+step1 + 2*step2)*(max_dims=data.dim // 5)`.

Reply:
Yes, please stay consistent to the baseline in terms of number of masked channels. Otherwise, it is not clear what causes the effects that we observe in the results.

Reply:
Solved by removing the custom function and adding a flag for shuffling